lib/extras/packed_image.h

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#ifndef LIB_EXTRAS_PACKED_IMAGE_H_
+#define LIB_EXTRAS_PACKED_IMAGE_H_
+
+// Helper class for storing external (int or float, interleaved) images. This is
+// the common format used by other libraries and in the libjxl API.
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "jxl/codestream_header.h"
+#include "jxl/types.h"
+#include "lib/jxl/base/status.h"
+#include "lib/jxl/common.h"
+
+namespace jxl {
+namespace extras {
+
+// Class representing an interleaved image with a bunch of channels.
+class PackedImage {
+ public:
+  PackedImage(size_t xsize, size_t ysize, const JxlPixelFormat& format)
+      : PackedImage(xsize, ysize, format, CalcStride(format, xsize)) {}
+  PackedImage(size_t xsize, size_t ysize, const JxlPixelFormat& format,
+              size_t stride)
+      : xsize(xsize),
+        ysize(ysize),
+        stride(stride),
+        format(format),
+        pixels(malloc(ysize * stride), free),
+        pixels_size(ysize * stride) {}
+  // Construct the image using the passed pixel buffer. The buffer is owned by
+  // this object and released with free().
+  PackedImage(size_t xsize, size_t ysize, const JxlPixelFormat& format,
+              void* pixels, size_t pixels_size)
+      : xsize(xsize),
+        ysize(ysize),
+        stride(CalcStride(format, xsize)),
+        format(format),
+        pixels(pixels, free),
+        pixels_size(pixels_size) {
+    JXL_ASSERT(pixels_size >= stride * ysize);
+  }
+
+  PackedImage& operator=(PackedImage& other) = delete;
+  PackedImage& operator=(PackedImage&& other) = default;
+
+  PackedImage(PackedImage& other) = delete;
+  PackedImage(PackedImage&& other) = default;
+
+  // The image size in pixels.
+  size_t xsize;
+  size_t ysize;
+
+  // Whether the y coordinate is flipped (y=0 is the last row).
+  bool flipped_y = false;
+
+  // The number of bytes per row.
+  size_t stride;
+
+  // The interleaved pixel as defined in the storage format.
+  JxlPixelFormat format;
+  std::unique_ptr<void, decltype(free)*> pixels;
```

Comment:
I'm not quite sure how happy I feel about this being public...

Reply:
ok, I made it confidential.